src/tasks/gitea/gitea-drone-oauth.ts

```diff
@@ -0,0 +1,225 @@
+
+  return authorizeResponse.headers['set-cookie']
+}
+
+export function parseCSRFToken(authorizeHeaderCookies: string[]): string {
+  // Loop over cookies and find the _csrf cookie and retrieve the value
+  const csrfCookies = authorizeHeaderCookies.filter((cookie) => {
+    return cookie.includes(csrfCookieName) // Find cookie that contains _csrf
+  })
+
+  if (csrfCookies.length == 0) {
+    throw new GiteaDroneError('No CSRF cookie was returned')
+  }
+
+  const csrfTokens = csrfCookies[0]
+    .split(';') // Split that cookie into multiple parts, as a single cookie can contain multiple key/values ';' separated
+    .map((c) => c.trim())
+    .filter((cookie) => {
+      return cookie.includes(csrfCookieName) // Find the key/value pair with _csrf as key
+    })
+    .map((cookie) => {
+      return cookie.substring(csrfCookieName.length + 1) // Retrieve value for '_csrf'-key
+    })
+
+  if (csrfTokens.length == 0 || csrfTokens[0].length == 0) {
+    throw new GiteaDroneError('No CSRF token was returned')
+  }
+  return csrfTokens[0]
+}
+async function authorizeOAuthApp(giteaUrl: string, droneLoginUrl: string, oauthData: DroneSecret) {
+  let authorizeHeaderCookies: string[]
+  try {
+    authorizeHeaderCookies = await getGiteaAuthorizationHeaderCookie(giteaUrl, droneLoginUrl, oauthData)
+  } catch (error) {
+    throw new GiteaDroneError('Authorization already granted or something went wrong')
+  }
+
+  const csrfToken = parseCSRFToken(authorizeHeaderCookies)
+
+  console.log('Granting authorization')
+
+  const grantOptions: AxiosRequestConfig = {
+    method: 'POST',
+    url: `${giteaUrl}/login/oauth/grant`,
+    headers: {
+      cookie: authorizeHeaderCookies
+        .map((cookie) => {
+          return cookie.split(';')[0]
+        })
+        .join('; '),
+      Cookie: authorizeHeaderCookies
+        .map((cookie) => {
+          return cookie.split(';')[0]
+        })
+        .join('; '),
+    },
+    maxRedirects: 1,
+    auth: {
+      username: env.GITEA_USER,
+      password: env.GITEA_PASSWORD,
+    },
+    // Data for this post query must be stringified https://github.com/axios/axios#using-applicationx-www-form-urlencoded-format
+    data: querystring.stringify({
+      _csrf: csrfToken,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      client_id: `${oauthData.clientId}`,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      redirect_uri: droneLoginUrl,
+    }),
+  }
+
+  try {
+    await axios.request(grantOptions)
+  } catch (error) {
+    // Do nothing, error code could be on the redirect
+  }
+}
+
+export function isSecretValid(remoteSecret: DroneSecret, oauthApps: OAuth2Application[]): boolean {
+  if (!remoteSecret) {
+    console.log('Remote secret was not found')
+  } else if (remoteSecret.clientId.length == 0 || remoteSecret.clientSecret.length == 0) {
+    console.log('Remote secret values were empty')
```

Comment:
But it is not an error. I don't mind changing it, but shouldn't the program stop on an error?
Maybe a `console.warn`?

Because if this if/elseif block fails, it will remove & recreate the secrets, so it is not breaking.

Reply:
The thing is we want to now allow the outside world to provide invalid input. So we just bork

Reply:
The outside world doesn't provide invalid input. We retrieve values, that we should have generated. But if we don't check to make sure the `remoteSecret` is filled. Because if we don't error out there it will try to authorise on an empty secret - and that's not possible. By having this check, we make sure that the secret in k8s is filled.
Otherwise, we will delete this secret, and all gitea oauth apps with the same name. And create fresh ones.

Reply:
If we determine that the preconditions have not been met we should just fail.

Reply:
No we don't need to fail, as we can recover from it. As explained in my comment above:

If data matches: Great
If data doesn't match: Remove old data, and re-create.